[Bug 18646] Use object handles instead of global object pointers (8.1)

livecodefraser opened the pull request:
Backport of https://github.com/livecode/livecode/pull/4762

livecodefraser committed:
Correct a comment about MCObjectHandles

A "not" was missing, completely reversing the meaning of the comment!

livecodefraser committed:
Change MCButton::menu to an MCStackHandle

Instead of MCObjectHandle - this removes lots of needless casts.

livecodefraser committed:
Make MCObjectHandle's MCObject* constructor explicit

Having this non-explicit makes various comparisons ambiguous.

livecodefraser committed:
Add comparisons between MC...Handles

livecodefraser committed:
Tweak object.h to not depend on globals.h

This allows globals.h to include object.h (needed in order to turn
various global variables into handles instead of raw pointers).

livecodefraser committed:
object.h no longer includes globals.h

Various other files needed fixed due to implicit dependencies.

livecodefraser committed:
Don't use raw object ptrs in the event queue

In case the object gets deleted while there is a queued event.

livecodefraser committed:
Turn MCtopstackptr into an object handle

livecodefraser committed:
Turn MCdefaultstackptr into an object handle

livecodefraser committed:
Turn MCstaticdefaultstackptr into an object handle

livecodefraser committed:
Turn MCmousestackptr into an object handle

livecodefraser committed:
Turn MCclickstackptr into an object handle

livecodefraser committed:
Turn MCfocusedstackptr into an object handle

livecodefraser committed:
Turn MCdynamiccard into an object handle

livecodefraser committed:
Turn MCmenuobjectptr into an object handle

livecodefraser committed:
Turn MCerrorptr into an object handle

livecodefraser committed:
Turn MCerrorlockptr into an object handle

livecodefraser committed:
Turn MCdefaultmenubar into an object handle

livecodefraser committed:
Turn MCplayers into an object handle

livecodefraser committed:
Turn MCacptr into an object handle

livecodefraser committed:
Turn MCmagimage into an object handle

livecodefraser committed:
Turn MCmagnifier into an object handle

livecodefraser committed:
Turn MCdragsource into an object handle

livecodefraser committed:
Turn MCdragdest into an object handle

livecodefraser committed:
Turn MCclickfield into an object handle

livecodefraser committed:
Turn MCfoundfield into an object handle

livecodefraser committed:
Turn MCdropfield into an object handle

livecodefraser committed:
Turn MCeditingimage into an object handle

livecodefraser committed:
Turn MCtooltip into an object handle

livecodefraser committed:
Turn MCmbstackptr into an object handle

livecodefraser committed:
Turn MCdragtargetptr into an object handle

livecodefraser committed:
Turn MCtracestackptr into an object handle

livecodefraser committed:
Use object handles for breakpoints and watched vars

livecodefraser committed:
Convert the ID cache to use object handles

livecodefraser committed:
Fix a very weird use of operators

...I don't really know what was going on there!

livecodefraser committed:
Fix various Linux-specific compile errors introduced by object handles

livecodefraser committed:
Fix a friend declaration to make GCC happy

Clang seems to be a lot laxer about friend declarations.

livecodefraser committed:
Don't call MCStack::domess on a nil pointer

MCdefaultstack hasn't been initialised at the time the Linux printer
code runs some LCS to parse some text. (I have no idea why this was
working before...). It now uses the template stack instead.

livecodefraser committed:
Fix an invalid object handle dereference in the breakpoint code

livecodefraser committed:
Fix the mobile-specific event code after MCObjectHandle refactor

The event code for touch events is only compiled on mobile platforms and
was missed in the first refactoring pass.

livecodefraser committed:
Fix Android-specific code broken by MCObject* replacement

livecodefraser committed:
Fix iOS-specific code broken by MCObject* replacement

peter-b commented on engine/src/eventqueue.cpp:
```diff
@@ -650,6 +636,33 @@ static MCStack *MCEventQueueGetEventStack(MCEvent *p_event)
+    else if (p_event -> type == kMCEventTypeUpdateMenu)
+    {
+        MCObjectHandle t_handle = p_event->menu.target;
+        t_handle.ExternalRelease();
```
So, obviously we need to construct an `MCObjectHandle` from the `target` pointer here so that it can be properly released.  However, it's not actually obvious that an `MCObjectHandle` _needs_ to be constructed here just by inspection.  So what about doing something like:

``` C++
MCObjectHandle(p_event->menu.target).ExternalRelease();
```

This would make it clearer that it's actually pretty important to construct the `MCObjectHandle`.

Obviously this code is correct as it is, this is just a suggestion to make it more "obviously correct".

peter-b commented on libfoundation/include/foundation-auto.h:
```diff
@@ -87,9 +87,14 @@ template<typename T> class MCAutoValueRefBase
+    
+    inline bool operator ! ()
+    {
+		return m_value == nil;
```
@livecodefraser I don't think we want to do this, because it introduces confusion when applied to an `MCAutoBooleanRef`.  For now, could you please leave `operator bool()` and `operator !()` undefined for `MCAutoValueRefBase` and add an `IsSet()` method (or equivalent) instead?

peter-b commented on engine/src/sellst.cpp:
```diff
@@ -98,30 +119,59 @@ bool MCSellist::getids(MCListRef& r_list)
 void MCSellist::top(MCObject *objptr)
 {
-	if (objects != NULL)
+	Clean();
```
@livecodefraser Could you please refactor this patch so that the list is only cleaned when the list is being mutated anyway?

peter-b reviewed (changes requested):
@livecodefraser This is an awesome technical accomplishment -- I'm amazed by how _few_ line changes were needed to make such a large change to the engine!

This looks pretty near ready to merge -- the only thing that I think is definitely not okay are the bool operators on `MCAutoValueRefBase`.

Thanks again for dealing with this big project.

livecodefraser committed:
Refactor MCObjectHandle to allow MCButtonHandle, etc

Due to compiler limitations, classes like MCStackHandle won't convert
to MCObjectHandle naturally: this can be done in a type-safe manner
using some features in C++11 that are not available on all of our
supported platforms at this time.

Conflicts:
	engine/src/exec-interface2.cpp
	engine/src/object.cpp
	engine/src/object.h

livecodefraser committed:
Update the friend declaration for MCMixinObjectHandle

GCC doesn't like the construction previously used (but Clang does) so
it has been updated to one that is acceptable to both.

Conflicts:
	engine/src/object.h

livecodefraser committed:
Make MCObjectHandle slightly less strict

Get() now no longer asserts that the handle is valid; it instead checks
that it is either unbound or valid. This change has been made to support
a common use pattern in the engine. It should not affect safety as
dereferencing an unbound pointer will cause a segfault.

N.B this pattern is passing nil to methods to indicate no object

Conflicts:
	engine/src/object.h

livecodefraser committed:
Turn MCObject::parent into an MCObjectHandle

This helps catch any erroneous accesses to parents that have been
deleted.

Conflicts:
	engine/src/exec-interface2.cpp
	engine/src/styledtext.cpp

livecodefraser committed:
Use ObjectHandles for the recent cards list

Conflicts:
	engine/src/cardlst.cpp

livecodefraser committed:
Turn MCmenubar into an object handle

Conflicts:
	engine/src/stack2.cpp

livecodefraser committed:
Turn MCactivefield into an object handle

Conflicts:
	engine/src/stack2.cpp

livecodefraser committed:
Turn MCactiveimage into an object handle

Conflicts:
	engine/src/stack.cpp

livecodefraser committed:
Turn MCtracedobject into an object handle

Conflicts:
	engine/src/debug.cpp

livecodefraser committed:
Fix an invalid pointer cast

Conflicts:
	engine/src/styledtext.cpp

livecodefraser committed:
Refactor the sel object list to use object handles

Conflicts:
	engine/src/sellst.cpp

peter-b commented:
@livecode-vulcan review ok 8911085

livecode-vulcan commented:
💙 review by @peter-b ok 8911085962c788ebb73eebc13051d17b88ae903e

livecode-vulcan commented:
😞 test failure 8911085962c788ebb73eebc13051d17b88ae903e
- try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/513)
- try-community-mac: [failure](https://vulcan.livecode.com/builders/try-community-mac/builds/518)
- try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/513)
- try-community-emscripten: [failure](https://vulcan.livecode.com/builders/try-community-emscripten/builds/517)
- try-community-ios: [pending](https://vulcan.livecode.com/builders/try-community-ios/builds/520)
- try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/513)
- try-community-android-armv6: [pending](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/514)

livecodefraser committed:
Fix various files failing to include globals.h

They were previously depending on it being implicitly included via
another header.

livecodefraser committed:
Fix an unchecked dereference of MCactivefield on OSX

livecodefraser committed:
Fix a comparison between bool and ES_NORMAL

livecodefraser committed:
Ensure object handle invalidation happens at the end of the destructor

Otherwise things that might need to make reference through them will fail.

livecodefraser committed:
Add an IsBoundTo method to MC<...>Handle

This acts a bit like operator== but doesn't fail if the handle has been
invalidated.

livecodefraser committed:
Fix various issues with the stack ID cache code

Re-defining integer limit macros is bad...

livecodefraser committed:
Fix the redraw code assuming only groups can be control parents

Things like the combobox have a field inside a button.

livecodefraser committed:
Turn an unchecked object cast into a checked cast

peter-b commented:
@livecode-vulcan review ok a5298d5

livecode-vulcan commented:
💙 review by @peter-b ok a5298d5fa4ff99c3eb37904c859edfcf08b2d124

livecode-vulcan commented:
😞 test failure a5298d5fa4ff99c3eb37904c859edfcf08b2d124
- try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/561)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/566)
- try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/561)
- try-community-emscripten: [failure](https://vulcan.livecode.com/builders/try-community-emscripten/builds/565)
- try-community-ios: [pending](https://vulcan.livecode.com/builders/try-community-ios/builds/570)
- try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/562)
- try-community-android-armv6: [pending](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/562)

livecodefraser committed:
Ensure <Windows.h> is only included via prefix.h

This means that the appropriate macros are defined and undefined everywhere
the file is included. (This is vital as there are some name clashes between
the Win32 headers and LiveCode method names, like GetObject).

livecodefraser committed:
Fix some uses of object handles in Win32-specific code

livecodefraser committed:
Add some ifdefs to make friend declaration work on all compilers

MSVC and GCC seem to have mutually-incompatible ways of requesting that a
template class or method is a friend. Similarly, GCC requires 'typename' in
contexts where MSVC rejects it.

livecodefraser committed:
Ensure em-stack.cpp includes globals.h

It previously relied on it being included by object.h

peter-b commented:
@livecode-vulcan review ok 129372e

livecode-vulcan commented:
💙 review by @peter-b ok 129372e6ec7336c1f3cb2b494de7498ae9fe5035

livecode-vulcan commented:
😞 test failure 129372e6ec7336c1f3cb2b494de7498ae9fe5035
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/562)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/567)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/562)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/566)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/571)
- try-community-win-x86: [failure](https://vulcan.livecode.com/builders/try-community-win-x86/builds/563)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/563)

livecodefraser committed:
Fix a clash due to the Win32 SIZE typedef in the C++ tests

peter-b commented:
@livecode-vulcan review ok f9b3687

livecode-vulcan commented:
💙 review by @peter-b ok f9b36870c811657c36b90924ce338564d5326135

livecode-vulcan commented:
😎 test success f9b36870c811657c36b90924ce338564d5326135
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/563)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/568)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/563)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/567)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/572)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/564)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/564)
